### PR TITLE
allow specifying multiple output file groups for binarize, OCR-D/spec#117

### PR DESCRIPTION
**Pull request by kba**
Proof of Concept how multiple output file groups can be supported.

### CODE REVIEW

**Comment by bertsky** [comment]
Sorry about the failing tests, that was a fault of mine from the last PR (simplify-common). Fixing now.

**Comment by bertsky** [comment]
@kba probably no need to rebase though – we do not even have binarization in the tests yet.

File: ocrd_tesserocr/binarize.py [hunk 1]
```diff
@@ -52,10 +52,15 @@ def process(self):
         
         Produce a new output file by serialising the resulting hierarchy.
         """
+        try:
+            self.page_grp, self.image_grp = self.output_file_grp.split(',')
+        except ValueError:
+            self.page_grp = self.output_file_grp
+            self.image_grp = FILEGRP_IMG
```
Comment (bertsky) [inline ocrd_tesserocr/binarize.py]: I really like this fallback. But we could give a warning here that images are going there.

**Comment by bertsky** [comment]
Also, what do we do with `output_file_grp` in `ocrd-tool.json`? There needs to be a place where assumptions on input and output file groups can be put (how many, what order)... same for multi-OCR input or OCR-GT alignment/evaluation

File: ocrd_tesserocr/binarize.py (hunk 1 above)
Comment (kba) [reply ocrd_tesserocr/binarize.py]: Done in 6c0ee6d

**Comment by kba** [comment]
> Also, what do we do with output_file_grp in ocrd-tool.json? There needs to be a place where assumptions on input and output file groups can be put (how many, what order)... same for multi-OCR input or OCR-GT alignment/evaluation

Tricky because not only can there be multiple output file groups for different data types where order is important, but also different "variants" depending on the input.

Maybe change the structure of `output_file_grp` to a more expressive array of objects, e.g.:

```yaml
"ocrd-tesserocr-binarize": {
  "executable": "ocrd-tesserocr-binarize",
  "categories": ["Image preprocessing"],
  "description": "Binarize pages, regions or lines",
  "input_file_grp": [
    "OCR-D-IMG",
    "OCR-D-SEG-BLOCK",
    "OCR-D-SEG-LINE"
  ],
  "output_file_grp": [
    {"description": "PAGE-XML output", "examples": ["OCR-D-BIN-BLOCK", "OCR-D-BIN-LINE"], "required": true},
    {"description": "Binarized images", "default": "OCR-D-IMG-BIN"}
  ],
  "steps": ["preprocessing/optimization/binarization"],
  "parameters": {
    "operation_level": {
      "type": "string",
      "enum": ["region", "line"],
      "default": "region",
      "description": "PAGE XML hierarchy level to operate on"
    }
  }
}
```

**Comment by bertsky** [comment]
Looks good already (regardless of the outcome of OCR-D/spec#117). I'd say somebody should systematically make this change in all `AlternativeImage` aware processors (Tesseract and Ocropy) now – before users and components get used to the current fixed scheme. (But that would not be me, sorry.)

**Comment by kba** [comment]
I can do that, but can you open an issue and assign me so i won't forget please? Thanks

**Comment by bertsky** [comment]
> I can do that, but can you open an issue and assign me so i won't forget please? Thanks

done – see above. One for Tesseract, one for Ocropy.

**Comment by wrznr** [comment]
@kba Can we merge?

**Pull request merged**